src/libraries/Common/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -127,67 +147,78 @@ public override bool TrySignHash(ReadOnlySpan<byte> hash, Span<byte> destination
                     return false;
                 }
 
-                ReadOnlySpan<byte> derSignature = SignHash(hash, signDestination, signatureLength, key);
-                bytesWritten = AsymmetricAlgorithmHelpers.ConvertDerToIeee1363(derSignature, KeySize, destination);
-                Debug.Assert(bytesWritten == encodedSize);
+                using (SafeEvpPKeyCtxHandle ctx = CreateCtx())
+                {
+                    ctx.ConfigureForECDSASign();
+
+                    if (!ctx.TryGetSufficientSignatureSizeInBytesCore(hash, out int sufficientSignatureSizeInBytes))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    Span<byte> derSignatureDestination = sufficientSignatureSizeInBytes <= SignatureStackBufSize ? stackalloc byte[sufficientSignatureSizeInBytes] : new byte[sufficientSignatureSizeInBytes];
+                    if (!ctx.TrySignHashCore(hash, derSignatureDestination, out int derSignatureBytesWritten))
```

Comment:
Ideally, we can drop SafeEvpPkeyCtx and just re-shim things to take the EVP_PKEY*, make a CTX, set it up, call sign, and exit.

If we really need different errors for the different phases, or there's some other reason that we need to be so chatty across the P/Invoke boundaries, I'd like to understand why.

Reply:
Is there any reason to not cross P/invoke boundaries? Is that slow for any particular reason? I'd prefer we move as much code to managed so I'm a bit confused on why is crossing boundary bad?

I'll address the "different phases errors" later - I can't remember why I did the split in the first place - I'll do some experimentations once I fix the build errors

Reply:
> Is there any reason to not cross P/invoke boundaries? Is that slow for any particular reason?

P/Invokes are expensive.  For SafeHandle there's the DangerousAdd/DangerousRelease wrapper around it, and for all calls there is always some GC handshaking.  They're not "never do them" expensive, but they are a "justify this" level of expensive, which means any time something can be done in one you need to justify making it two.

> I'd prefer we move as much code to managed

Code needs to live at the right layer for a combination of correctness and maintainability.  Moving code into managed doesn't inherently make it better (see the UAF this change tried to introduce).

The idealized layers look like:

* (Normal-)User code: Lives happily above the layer of caring about what OS they're on, what kind of CPU it is, the endianness of said CPU, whether time_t is 32 or 64 bits (for the next little while, at least).
* RSA.cs: Models RSA for .NET.  Doesn't have any SafeHandles, definitely doesn't have any pointers tracked as `nint`/`IntPtr`/`T*`, doesn't do any crypto-math, just sets up some public functions, maybe some overloads to those functions, and empowers specific implementations to do the real work.
* RSAOpenSsl.cs/RSAAndroid.cs/etc: Translates the abstractions into terms that the back-end provider understands. Manages the lifetime relationship between the abstract concept and the back-end.
  * Example, Decrypt:
    * The .NET API says the Span-writing Decrypt will succeed if the decrypted content fits into the destination span, but the underlying provider thinks differently, requiring a bigger minimum size. OK, so arrange for a temporary buffer.
      * How big it needed to be, whether it's done always or just when destination is too small, etc, are things tracked by, and managed by, this layer.
    * The hash algorithm and padding mode need to be translated from the .NET versions to the provider versions.  Maybe that's "call different functions", maybe that's "use a lookup table" (and maybe it's even "we'll send the .NET values on to a later level").
    * "Sign stuff".
    * Turn the back-end answer into the answer from this routine.
* Interop.Crypto.*: Translate the intent between "we want to call a function in the back-end" and doing so.
  * Exposes functions that accept Span/ReadOnlySpan, expands them into the pointer and length to call non-exposed functions.
  * If this function was purely managed, should it be `void`?  OK, then trap the error message and turn it into a throw here, so the caller can't forget to.
    * Similarly if the native gives back an answer and a success/failure response, trap the error state here, don't leak it out.
* pal_evp_pkey.c / android_stuff.c / android_stuff.java / macOS_stuff.c / macOS_stuff.swift / etc: Expose a function that can be called by .NET (via Interop.*) that achieves a task.
  * If state is involved (probably a SafeHandle on the .NET side) then this should strive to be an atomic operation, to avoid dealing with questions about torn state.


OpenSSL previously had a nice atomic function, ECDSA_sign.  You call it with an `ECDSA*`, it internally figures out what it means to do ECDSA math with the key you're tracking, and it does the math using the other inputs, and writes it to the destination you gave it.  We probably called that in an era when all we did in Interop was define `internal extern int ...`, and left work that now lives in that layer to the caller... but if we wanted to call it today it'd be something like

Interop:
```C#
[LibraryImport(...)]
private static unsafe partial int ECDsaSign(SafeECDsaHandle key, byte* hash, int hashLen, byte* destination, int destinationLength, out int bytesWritten);

internal static unsafe int ECDsaSign(SafeECDsaHandle key, ReadOnlySpan<byte> hash, Span<byte> destination)
{
    int ret;
    int written;

    fixed (byte* hashPtr = hash)
    fixed (byte* destPtr = destination)
    {
        ret = ECDsaSign(key, hashPtr, hash.Length, destPtr, destination.Length, out written);
    }

    if (ret != 1)
    {
        throw CreateException();
    }

    return written;
}
```

ECDsaOpenSsl.cs:

```C#
public int SignHash(ReadOnlySpan<byte> hash, Span<byte> destination)
{
    // Assume this checks for disposed, etc.
    // Why do we want a local here? In case we need to call more than one P/Invoke and need to be sure the key never changed.
    // What happens if it parallel disposes? Probably an exception.  But not an OOB write, or anything else scary.
    SafeECDsaHandle key = GetKey();

    if (hash.IsEmpty)
    {
        throw something;
    }

    // If no parallel activity happens, this gives a nice exception.
    // If we just raced with importing a bigger key, we'll not throw here, but it'll get caught later.
    if (destination.Length < _minSignatureLength)
    {
        throw something;
    }

    return Interop.Crypto.ECDsaSign(key, hash, destination);
}
```

and the backend:

```C
PALEXPORT int32_t CryptoNative_ECDsaSign(
    SomeECDSA* key,
    uint8_t* hash,
    int32_t hashLen,
    uint8_t* dest,
    int32_t destLen,
    int32_t* bytesWritten)
{
    assert(key && hash && hashLen && dest && destLen && bytesWritten);
    // We said it's a .NET "out", so always have to write.
    *bytesWritten = 0;

    // Hey, ECDSA_sign doesn't take destLen, that means it'll do an OOB-write, so do one
    // last check for the size.

    int minSize = ECDSA_min_buf_size(key);

    if (destLen < minSize)
    {
        return a special error code, or ERR_push_error a thing and return general failure.
    }

    int ret = ECDSA_sign(key, hash, hashLen, dest);

    if (ret > 0)
    {
        *bytesWritten = ret;
        return 1;
    }

    return 0;
}
```

The native function could be smaller if we moved the length check against that specific key up into managed... but is it worth it? We're doing the exact same amount of work, but now it's less guaranteed.  We make the native code _less_ safe by moving that piece.

OK... so now what about when they change from SomeECDSAKey* to EVP_PKEY*?  Well... just change the one thing in our PALEXPORT, and then make that function continue doing this one atomic operation.  Then we're never calling EVP_PKEY_sign with a CTX that isn't in sign-mode, we're still guarding against parallel key imports, etc.  And, we have a very, very clear "how do we sign a key in ECDSA" for when some poor schmuck (me) has to write that down in the cryptographic disclosures document :smile:.

(And, yes, I skipped over the IEEE P-1363 vs X.509 DER signature format... that would belong in ECDsaOpenSsl.cs, since it's managing the relationship between the backend signature format and the .NET API)